### ISSUE TITLE
tf.io.gfile.glob does not list all files in a Google Cloud Storage bucket

### ISSUE THREAD

**Issue by SaschaHeyer** [tensorflow/tensorflow#34371]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): ?
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: /
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0.0
- Python version: 3
- Bazel version (if compiling from source): /
- GCC/Compiler version (if compiling from source): /
- CUDA/cuDNN version: /
- GPU model and memory: /

**Describe the current behavior**
When listing file with `tf.io.gfile.glob` not all images are returned. It seems it is not resolving the folders recursively. 
When using the same path with gsutils  we get the correct image count.

**Describe the expected behavior**
When using the same gs:// path with **gsutils** we get the correct amount of images. 

**Code to reproduce the issue**
In order to reproduce the behavior I prepared a Google Bucket with the following structure. The bucket is public accessible, please feel free to use it to reproduce the behavior on your end: `gs://tensorflow-issue-reproduction`

![level0](https://user-images.githubusercontent.com/1991664/69040135-93375e80-09ed-11ea-927f-da445d2bae10.png)
![level1](https://user-images.githubusercontent.com/1991664/69040139-94688b80-09ed-11ea-9e19-28af47c3bcee.png)
![level2](https://user-images.githubusercontent.com/1991664/69040146-96cae580-09ed-11ea-8894-c69a82acbbaa.png)
![level3](https://user-images.githubusercontent.com/1991664/69040150-9894a900-09ed-11ea-94dc-49fb2d9b3b9b.png)

In summary we have 4 jpg images nested in different folder levels.

TensorFlow 2 code to reproduce
```
files = tf.io.gfile.glob('gs://tensorflow-issue-reproduction/**/*.jpg')
print('file count: ', len(files))
# found files 1
```

gsutil command which works properly
```
gsutil du gs://tensorflow-issue-reproduction/**/*.jpg | wc -l
# found files 4
```

**Other info / logs**
/

Best regards
Sascha

**Comment by SaschaHeyer** [comment]
Hi @ymodak 

any ideas?

Best regards
Sascha

**Comment by mihaimaruseac** [comment]
Will take a while until I can get to this, but it's something to be handled as part of https://github.com/tensorflow/community/pull/101

**Comment by SaschaHeyer** [comment]
Hi @mihaimaruseac 

what does this mean "as part of tensorflow/community#101"?
What are the next steps that are required to solve this issue?
Who is in charge?

**Comment by mihaimaruseac** [comment]
It means we are refactoring all of the infrastructure around filesystem support. Currently POSIX is moved to the new world, Windows follows then GCS and the others. I am coordinating this process

**Comment by kureta** [comment]
This issue still persists. Am I missing something? Is recursive globbing support dropped or something like that? Thanks.

**Comment by mihaimaruseac** [comment]
Unfortunately we didn't yet get to this issue. Apologies for the delay

**Comment by kureta** [comment]
@mihaimaruseac thank you very much for the response. no worries.

**Comment by mihaimaruseac** [comment]
See also, #39565

**Comment by mihaimaruseac** [comment]
I just noticed you are using `**` but TF's [globbing patterns don't include that case](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/platform/file_system.h;l=184-215;drc=516ae286f6cc796e646d14671d94959b129130a4)

Including `**` would complicate [an already complicated implementation](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/platform/file_system_helper.cc;l=123-266;drc=8b5b9dc96666a3a5d27fad7179ff215e3b74b67c) but I think we can take in PRs to add support for `**`

**Comment by sushreebarsa** [comment]
@SaschaHeyer Could you please refer to the above comment and let us know if you still face this issue ?Thanks!

**Comment by wookayin** [comment]
Was there any relevant changes in `tf.io.gfile` to change the behavior?

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by mihaimaruseac** [comment]
Removing the stale label since this has a clear reproduction and there has been no work to actually fix it.

**Comment by Venkat6871** [comment]
Hi @SaschaHeyer  ,
Thank you so much for filing this bug request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.To avoid duplicate work, please leave a comment here if you decide to take it on!
Thank you again for your interest and we look forward to seeing your contributions!
